File: cpp/allergies/allergies.cpp
#include "allergies.h"

allergies::allergy_test::allergy_test(int allergy_score) {
  std::bitset<8> allergy_bitset(allergy_score);

  // Check bitset positions.
  if (allergy_bitset.test(0)) allergies.insert("eggs");
  if (allergy_bitset.test(1)) allergies.insert("peanuts");
  if (allergy_bitset.test(2)) allergies.insert("shellfish");
  if (allergy_bitset.test(3)) allergies.insert("strawberries");
  if (allergy_bitset.test(4)) allergies.insert("tomatoes");
  if (allergy_bitset.test(5)) allergies.insert("chocolate");
  if (allergy_bitset.test(6)) allergies.insert("pollen");
  if (allergy_bitset.test(7)) allergies.insert("cats");

}

bool allergies::allergy_test::is_allergic_to(std::string allergen) {
  return allergies.find(allergen) != allergies.end();
}

File: javascript/allergies/allergies.js
const ALLERGIES = [
  'eggs',
  'peanuts',
  'shellfish',
  'strawberries',
  'tomatoes',
  'chocolate',
  'pollen',
  'cats'
];

export class Allergies {
  constructor(num) {
    const allergySelectorArray = (num % 256)
      .toString(2)
      .split('')
      .reverse()
      .map(Number);

    this.allergic = ALLERGIES
      .filter((_, i) => allergySelectorArray[i] === 1);
  }

  list() {
    return this.allergic;
  }

  allergicTo(item) {
    return this.allergic.indexOf(item) !== -1;
  }
}
